### PR TITLE
Move WebPreview to calypso-ui package

### LINKED ISSUE #34295: WebPreview component should be exported in packages
It will be great if we can reuse the `WebPreview` component in other projects.  In order to do this, we'll need to move the existing component from the `components/` folder to `packages/calypso-ui`.

In order to do this, we'll probably need to extract the following props/functions inside the component:

* Translation - currently uses the passed in `translate` prop.  Maybe we can use `@wordpress/i18n` instead?
* HoC - Remove or wrap any HoC `connect` wrappers separately in Calypso.
* State related items - `setPreviewShowing`, `isInlineHelpPopoverVisible`, and any others.

#### Screenshot / Video
Screen Shot 2019-06-26 at 1 23 29 PM

#### Context / Source
https://github.com/woocommerce/woocommerce-admin/issues/2469

### PR DESCRIPTION
**Pull request by joshuatf**
#### Changes proposed in this Pull Request

* Moves the `WebPreview` component and necessary dependencies to the `calypso-ui` package.
* Refactors the `WebPreview` component and separates device logic.
* Moves the `AsyncLoad` component to `calypso-ui`.
* Moves the `SpinnerLine` component to `calypso-ui`.

#### Thoughts
* There's one test still failing for translation.  Not sure if this has something to do with the default `translate` prop added here or another error.  Any feedback here appreciated!
* In order to keep the scope down on this PR, I chose to separate out the toolbar from the `WebPreview` component.  While some of this logic could benefit from being shared across projects, there are many dependencies tied up in Calypso.
* Similarly, the touch-detect `hasTouch()` method was manually copied over to avoid creating a new `calypso-lib` package.
* Devices and logic around the SEO panel were separated from the main component since these will probably vary widely from project to project.

#### Testing instructions

* Visit all the areas around the site that use the `WebPreview` component and make sure they continue to work as expected with correct buttons/content appearing based on props.
* Make sure that toggling the device in the toolbar continues to work as expected (especially when toggling between custom preview content types like `seo`).
* Make sure that the `SpinnerLine` continues to work across the site.
* Make sure the inline help popover does not lose focus if open before loading a new URL in the iframe.
* Optionally: `npm link` the calypso-ui package, bump the version, and test that you can import the `WebPreview` component in another project.

Fixes #34295

### CODE REVIEW

File: packages/calypso-ui/src/web-preview/content.jsx [hunk 1]
```diff
@@ -47,37 +48,39 @@ export class WebPreviewContent extends Component {
 	}
 
 	componentDidMount() {
+		const { onDeviceUpdate, previewMarkup, previewUrl } = this.props;
+
 		window.addEventListener( 'message', this.handleMessage );
-		if ( this.props.previewUrl ) {
-			this.setIframeUrl( this.props.previewUrl );
+		if ( previewUrl ) {
+			this.setIframeUrl( 'https://themes.woocommerce.com/?name=galleria' );
```
Comment (justinshreve) [inline packages/calypso-ui/src/web-preview/content.jsx]: This should be `previewUrl`.

Comment (joshuatf) [reply packages/calypso-ui/src/web-preview/content.jsx]: I guess we should not force every preview here to a WC theme 🤔  - updated!

File: packages/calypso-ui/src/web-preview/toolbar.jsx [hunk 2]
```diff
@@ -48,34 +50,37 @@ class PreviewToolbar extends Component {
...
-	constructor( props ) {
+	constructor() {
 		super();
 
 		this.devices = {
-			computer: { title: props.translate( 'Desktop' ), icon: 'computer' },
-			tablet: { title: props.translate( 'Tablet' ), icon: 'tablet' },
-			phone: { title: props.translate( 'Phone' ), icon: 'phone' },
-			seo: { title: props.translate( 'Search & Social' ), icon: 'globe' },
+			computer: { title: __( 'Desktop' ), icon: 'computer' },
```
[the remark this answers is not in the document]
Comment (joshuatf) [reply packages/calypso-ui/src/web-preview/toolbar.jsx]: @justinshreve Updated these, but still getting some errors with translation in tests.  Would you mind taking a look?

[the remark this answers is not in the document]
Comment (justinshreve) [reply packages/calypso-ui/src/web-preview/toolbar.jsx]: I did a little bit of digging, but I'm not sure why this is happening.

It looks like the [seo settings page](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-settings/seo-settings/form.jsx#L453-L461) that is failing uses the `WebPreview` component to render social previews, but I tested that feature in browser and it seems to be working properly.

I'm not sure why `translate` isn't getting passed though to the inline help text. Maybe someone more familiar with Calypso or that has worked on the code can help?

[the remark this answers is not in the document]
Comment (justinshreve) [reply packages/calypso-ui/src/web-preview/toolbar.jsx]: Per my Slack message, adding `translate: x => x,` to the `i18n-calypso` mock in `client/my-sites/site-settings/seo-settings/test/form.jsx` seems to fix the translate error.

However, now I am getting the following failure:

```
 FAIL  client/my-sites/site-settings/seo-settings/test/form.jsx
  Test suite failed to run

    ReferenceError: window is not defined

      at Object.<anonymous> (node_modules/component-event/index.js:1:39)
      at Object.<anonymous> (node_modules/click-outside/index.js:6:1)
```

It looks like `clickOutside` is used by `Popover`, which I assume is used by the inline help that the WebPreview can show.

I've been running into my own `window` issues over here with server side rendering: https://github.com/Automattic/wp-calypso/pull/34277#pullrequestreview-254432215.

**Comment by justinshreve** [comment]
It looks like my changes fixed the command line test, but we still have one failure in the desktop tests around image/svg handling.

https://circleci.com/gh/Automattic/wp-desktop/41728

I wonder if we need something like https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-build/jest.config.js and/or https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-build/webpack/file-loader.js for building the ui package?

**Comment by joshuatf** [comment]
It looks like the `calypso-build` should handle those file types.  Not quite sure why this isn't working for desktop.  Would appreciate input or guidance from anyone who better understands what's happening in the desktop build.

**Comment by jsnajdr** [comment]
Hi @joshuatf 👋 Thanks you for working on this component: it looks like a huge task 🙂 

The reason why the Desktop build fails is that the Desktop Server imports the Themes section (which is server-side-rendered) which in turns imports the Web Preview component, and the `SeoPreviewPane` imported from there imports a PNG image as a webpack asset. In order for that to work server-side, webpack needs to be configured to ignore images and other non-JS assets. And it's not configured that way for Desktop Server build.

That would be an easy patch, but the error reveals a more serious issue: async loading of the `WebPreview` is broken for some reason and as the [performance advisor comment](https://github.com/Automattic/wp-calypso/pull/34337#issuecomment-506312711) says, this PR increases the size of the entrypoint chunk by 25%. And that's a lot.

Can you look into that issue? None of the Web Preview code should be loaded until some preview is actually displayed. However, the current `components/web-preview` component directly imports `SeoPreviewPane` and `./toolbar`.

I think the best way to proceed is to start merging non-controversial parts of this big patch as separate PRs. For example, moving `SpinnerLine` to `calypso-ui` can be done independently. So can probably be done many of the refactorings.

This diff is really big, I see many weird and suspicious things when skimming through it, but I'm unable to do a comprehensive review of a patch of this size.

Some tips:
- `calypso-ui` components shouldn't use `AsyncLoad`. Code splitting decisions should be done in the app, not in a UI library
- `AsyncLoad` should be a separate package if you really need to extract it. It's a specialized tool, not a UI component. We'll eventually remove it completely in favor of a 3rd party library like [loadable-components](https://github.com/smooth-code/loadable-components)
- `RootChild` should be replaced by `React.createPortal` if possible. React didn't have a standardized solution yet when `RootChild` was created and we didn't migrate yet (see #32899 for the WIP of the migration attempt and why it's not trivial)

**Pull request closed without merging**